src/qcodes/dataset/dond/do_nd_utils.py

```diff
@@ -120,14 +120,22 @@ def _register_actions(
 
 
 @contextmanager
-def catch_interrupts() -> Iterator[Callable[[], MeasInterruptT]]:
-    interrupt_exception = None
+def catch_interrupts() -> Iterator[Callable[[], MeasInterruptT | None]]:
+    interrupt_exception: MeasInterruptT | None = None
+    interrupt_raised = False
 
-    def get_interrupt_exception() -> MeasInterruptT:
+    def get_interrupt_exception() -> MeasInterruptT | None:
         nonlocal interrupt_exception
         return interrupt_exception
 
     try:
         yield get_interrupt_exception
     except (KeyboardInterrupt, BreakConditionInterrupt) as e:
         interrupt_exception = e
+        interrupt_raised = True
```

Comment:
I don't think the `BreakConditionInterrupt` should bubble up. That one is intended to cancel a specific dond when a given condition is meet.